src/hardware/dma.cpp

```diff
@@ -100,29 +106,71 @@ static void perform_dma_io(const DMA_DIRECTION direction,
 	} while (remaining_bytes);
 }
 
-DmaChannel * GetDMAChannel(uint8_t chan) {
-	if (chan<4) {
-		/* channel on first DMA controller */
-		if (DmaControllers[0]) return DmaControllers[0]->GetChannel(chan);
-	} else if (chan<8) {
-		/* channel on second DMA controller */
-		if (DmaControllers[1]) return DmaControllers[1]->GetChannel(chan-4);
+void TANDYSOUND_ShutDown(Section* = nullptr);
+
+static bool activate_primary()
+{
+	assert(!primary);
+	constexpr uint8_t primary_index = 0;
+	primary = std::make_unique<DmaController>(primary_index);
+	return true;
+}
+
+static bool activate_secondary()
+{
+	// We need an AT architecture to support the secondary controller
+	if (!IS_EGAVGA_ARCH) {
+		LOG_WARNING("DMA: Cannot activate the secondary (high) DMA controller "
+		            "because the machine needs to be at least EGA");
```

Comment:
Would it be better to be a bit more precise here and say "because the machine type needs to be at least 'ega'"?

This is something new to me that we default to `sbpro2` on the `hercules` and `cga` machine types and only allow SB16 on `ega` or better. Is there a technical reason for this limitation? In real life, people were using Hercules as a secondary display adapter right until the end of the 90s, and you can use CGA and Hercules adapters in 386 and 486 systems without problems. Maybe Pentium too, but not sure about that.

I guess we might want to keep the defaulting this way due to backward compatibility reasons, but is it not possible to simply just allows any SB type on any machine type, excluding `pcjr` and `tandy` as those emulate some very specific hardware?

Also, starting DOSBox simply with `-noprimaryconf -machine hercules` now suddenly generates lots of warnings.

<img width="1178" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/71649d7a-813e-4f6e-9b32-3e4ac1d6e3a4">

This seems a little silly as `sbpro2` is what has been auto-selected based on the machine type, at least conceptually. What's really happening I think is that we default to `sb16` which then gets "demoted" to SB Pro2, and with the new logging we're informed about this in excruciating detail 😅 
 
If we really want to keep this behaviour, I think introducing `sbtype = auto` is in order perhaps?

Reply:
> This is something new to me that we default to sbpro2 on the hercules and cga machine types and only allow SB16 on ega or better. Is there a technical reason for this limitation? In real life, people were using Hercules as a secondary display adapter right until the end of the 90s, and you can use CGA and Hercules adapters in 386 and 486 systems without problems. Maybe Pentium too, but not sure about that.

Yup, the `machine` setting sets up the BIOS and motherboard memory, registers, and IO handlers to provide a period-correct environment for the give video device. 

`hercules`, `cga`, `tandy`, and `pcjr` were released when XT-class machines were only available, and software that was written in this pre-AT era could make assumptions (or confirm) that the machine meets the XT specifications.

Where as software written in the AT-era had to provide backward compatibility for XT systems and hardware; exactly like you said, people could plug in a CGA card to their pentium-class AT motherboard and BIOS.

For the best CGA experience, when a given game was written in the XT timeframe (and EGA wasn't even a thing then), using `machine = cga` gives the game the correct XT environment, and yeah - those motherboards on had a single 8-bit DMA controller.

Reply:
> This seems a little silly as sbpro2 is what has been auto-selected based on the machine type, at least conceptually. What's really happening I think is that we default to sb16 which then gets "demoted" to SB Pro2, and with the new logging we're informed about this in excruciating detail :sweat_smile:

Yes :-)  I'll knock down that logging :laughing: , and just mention that we're falling back to SB Pro2.

Reply:
> I think introducing `sbtype = auto` is in order perhaps? Then sbpro2 would be selected for anything lesser than ega and we can omit this downgrading dance...

That's a great suggestion!

- Perhaps older SB's might be preferred in the `cga`, `hercules` and `pcjr` machine types, so CMS is also available?

- Also, if a user is explicitly setting `machine = ega`, we /might/ want to give them `sbpro2`, because that's probably a better fit for the era of software they're using.  As an example: were there ever EGA-authored games that used 16-bit audio samples? EGA/<=286-era faded out as the primary target platform for games by ~1991 with devs now targeting VGA/386/486, and the SB16 wouldn't be released until 1992.

Also, the SB16 CT mixer's volume ramp is not backward compatible with pre-SB16 cards' volume ramp.  If you take a pre-SB16-era game that happens to adjust voice volumes and give it an SB16, then those volumes will be /way/ too quiet. They require the older volume ramp.  You'll see users working around this with the `sbmixer = false` setting.. but really, that's an indication the game wasn't authored for the SB16.

Reply:
> But I guess I'd prefer removing the limitation altogether except for Tandy/PCjr if possible (but we might still wanna keep the defaulting via the auto type).

The constraint here is that the XT machines didn't have a 16-bit DMA controller - and for those machine types, we really should keep it that way.

However, because this PR directly tests if the SB16's configured  `"hdma"` DMA channel is available, it is now possible to use an SB16 with these XT machine types.  You simply have to set `"hdma = 1"` or `"hdma = 3"`, which will use the 8-bit DMA controller to move 16-bit audio samples.

Edit: turns out this was possible with hardware too using the same technique (configuring the SB16's jumpers to use DMA address 1 for both the low and high channels):

### Running a Sound Blaster 16 on an XT

https://dosdays.co.uk/topics/sb16_on_old_systems.php

Reply:
> > But I guess I'd prefer removing the limitation altogether except for Tandy/PCjr if possible (but we might still wanna keep the defaulting via the auto type).
> 
> The constraint here is that the XT machines didn't have a 16-bit DMA controller - and for those machine types, we really should keep it that way.

Isn't that kind of moot, since there is no support for XT (or even AT) machine types (e.g. 8086/8088 or 286 CPUs) at this point? By default DOSBox emulates a 486 type CPU which will always have two PICs.

Reply:
I guess both be satisfied:

- Out of the box, `sbtype = auto` can respect the XT machine-types by picking an early SB model, similar to the "you asked for sb16 but we gave  you sbpro2 approach).

- But if the user demands a 16-bit DMA channel on an XT machine type ( `sbtype = sb16` or `gusdma = 6`), then we can still allow it.

Reply:
Very good points @kcgen, I like these ideas and would like them to be implemented. 🎉  Rather than replying to the individual points here, I started writing a proposal I'm going to post as a "discussion" soon because we're touching on some overarching concepts here that goes beyond this single enhancement.

Reply:
Okay, so I was quite enthusiastic about this auto-configuration mechanism and started writing a proposal about how we could configure all sorts of settings based on the `machine` type. Long story short, the more I thought about this, the more I disliked it. Something rubs me very much the wrong way with making all sorts of choices based on the *type of graphics adapter* (!) in the system.

_**Clippy, the DOSBox Assitant:** Alrighty, so because you've configured a `machine` with a Hercules graphics adapter, I'll set the `cycles` setting for you to XT speeds, moreover, I'll give you an `sb1` card, and maybe also disable `xms`, `ems`, and `umb`... Isn't that just W–O–N–D–E–R–F–U–L?_

Um, no, it's not. Just nope. Get lost.

The problem with such extensive implicit, invisible auto-config behaviour is that it can become very confusing and very hard to figure out what really is going on. For some people even the layered config approach is just too much, and frankly, I often confuse myself when I have my global config, local config, plus I also inject some further configs with the `-conf` option. You really need to know what you're doing if you're going down this route, and it's just too much for ordinary users.

Instead, I'm quite convinced providing some machine templates in the new user manual is the way forward that users can _copy_ into their local config and adjust further. That way everything is kept explicit, not some weird cascading mechanism where you have your global config, your local config, your "machine template config", maybe DOSBox also auto-configures a few things here and there... argh, my brain hurts just thinking about it (and the massive influx of confused people raising support tickets left, right, and center).

So, after so much tirade, what is the _least_ confusing, most explicit option we can do here? Do we _have_ to downgrade the SB16 or GUS to something else when a `machine` lesser than `ega` is configured? I'd much rather do the `hdma` configuration change you mentioned @kcgen than changing the `sbtype` by magic. If the config says `sb16`, we should do our absolute best to give the use `sb16` by whatever means necessary—especially because historically we default to `sb16` and SB is enabled by default. And yeah, I _really_ dislike the "shapeshifting Sound Blaster model" behaviour based on the *graphics adapter* type, let's not go down this path at all.

Reply:
I do think that having some sensible default "template" machine options for people to select between would be a good option, but that could come with the GUI, whenever that is. Something like:

- XT machine type with Hercules and SB 1.5
- XT machine type with CGA and SB 1.5
- AT machine type with EGA and SB Pro 2
- 386 machine type with VGA and SB Pro 2
- 486 machine type with SVGA and SB16
- Pentium machine type with SVGA and SB16
- Tandy with DAC machine type (286?)
- PCjr machine type

Reply:
Yeah we can revisit it later. Some GUI for this would be a viable option, because then you select a template which just sets various settings, but all settings are laid out before you on the GUI. You see exactly what settings the template set for you. Whereas if things get configured "magically" or by setting something like `configuration = 486` in the config, then how do you know what that actually configured for you? You have no idea, which makes further tweaking of those settings very difficult and confusing. Yes, we could document it, but if you have to look things up constantly, you're better off copying a template example from the manual into your local config, then it's all laid out before you for further tweaking... I don't think it gets better than that.

But yeah, a GUI approach takes all uncertainty out of that. You instantiate the template, you see exactly what the template configured for you, tweak further, then save. But the GUI approach would not play well at all with layered configs. They're almost totally mutually exclusive, at least to maintain them in a sane way. If a GUI was involved, I'd made sure that all layering and global config usage is turned off, and the GUI fully manages all the defaulting, templates, and whatnot, and always writes a _full_ local config, and even switches off global config loading with `-noprimaryconf`. How I see it, you either manually configure things, then you get layering, or you use a GUI, which then fully takes over the config management. Attempting to reconcile the two is a route to madness, I think.

For the record, the planned OSD is not really meant to be a full-blown DOSBox configurer/frontend sort of utility, at least in my mind. It would only allow changing settings which are useful to change in realtime while a game is running. Reconfiguring your whole machine that at the end requires a full emualtor restart anyway is out of scope, I think. A good guiding principle would be that if something cannot be tweaked at runtime while the game is running, then it won't belong to the OSD.

Reply:
We're in the DOSBox world, so we need to remember that `machine` is [supposed](url) be the entire system. 

What does this mean in practical terms?

- If you plugged a real TGA graphics card and TandSound card into a 286 DOS PC, then it still wouldn't run most Tandy-specific games. `machine = tandy` on the other hand will, [because it fakes out the BIOS chips, too](https://github.com/dosbox-staging/dosbox-staging/blob/main/src/ints/bios.cpp#L775:L808), and perhaps other details that I'm not aware of.

- `machine = tandy` is supposed to mimic an entire Tandy 1000 system so that all "APIs" (BIOS, IO devices, video, audio, etc) look like that machine, well, enough to coax the Tandy games to run. 

The point is that `machine = ` is not just swapping the graphics adapter in a bog-standard "PC clone".  It's meant to be the entire 'holistic' machine enough to satisfy picky games that really do need the `machine = ` specification.  (yes: the CPU model and speed doesn't follow suite: I suspect the DOSBox devs left that as-is because the _majority_ of games aren't CPU-sensitive and thus are unaffected by this)

> what is the least confusing, most explicit option we can do here? If the config says sb16, we should do our absolute best to give the use sb16 by whatever means necessary.

We should respect that - and give it to them, even if it means turning on the 16-bit DMA controller in those old machine types.

This PR also prevent DMA conflicts producing broken zombie sound cards.

So if SB wants DMA 1, and GUS wants DMA 1, and TandySound wants DMA 1 -> then only the last request will get it and prior users will be shudown explicitly in order, like musical chairs.  But we'll inform the user at each change.

> Do we have to downgrade the SB16 or GUS to something else when a machine lesser than ega is configured? 

I also think the magic downgrade "You set `sbtype = sb16` but we gave you `sbtype = sbpro2` -> that's asking for problems, especially if a user has configured their game explicitly for sb16.

Same with any magic changes to GUS *"You asked for `gusdma = 6`, but we sneakily configured `gusdma = 3`"* -> that's evil and also will break user's games that have hard-coded DMA values.

But! We do have a better opportunity with `sbtype = auto` to provide reasonable period "sane" SB types (especially critical to get the volume ramp correct), and for that, @rderooy's suggestion makes a lot of sense.

Reply:
Ok, we both think the magic downgrade is a bad idea, so far so good.

About the auto-config, I get it that the `machine` type is a fuzzy concept and it does not just set the graphics adapter. Arguably, this whole machine concept is strongest with `pcjr`and `tandy`--with the rest it's much looser.

One issue with `sbtype = auto` is that the choice of SB type is pretty arbitrary, there's no clear best setting.

- SB1 came out in 1989. With what SB adapter do you configure the "Hercules XT"? I'd say `none`. But maybe `gb`, `sb1`, or even `sbtype = none` and ` oplmode = opl2` are reasonable and realistic options.

- Exact same issue with CGA.

- PCjr and Tandy should get `none`, that's clear.

- With EGA it's clearer that at least OPL2 is reasonable, but probably SB1 or SB2 are best. EGA machines are Sierra machines 😎 

- With SVGA types do we differentiate? Or set a single SB model for all of them? 

At the end of the day, regardless that emulating a full Tandy and PCjr machine can requested via the `machine` setting, otherwise when I request `cga` or `ega` or °svga_s3`, I'm still thinking about graphics. My choice of graphics affecting what SoundBlaster model I get (including perhaphs none at all) is very counterintuitive and unexpected, and in many cases the auto-selected choice is pretty arbitrary. 

My other problem is, where do you draw the line? Also auto-default `cycles` based on `machine`, and `memsize` too? I started writing it down what to default these other settings to per each `machine`, I can share it with you. But then I realised, I won't remember what the auto-defaults per machine would be; and there won't be any form of good feedback on the auto-configured behaviour.

We can continue to play with these ideas for sure (I abandoned them in favour of explicitness), but what do you think about the two above issues? My ideal solution is user education: give them period accurate config examples in the manual, mentioning the possible variations for each, then they should use those as starting points and tweak further.

Reply:
> But! We do have a better opportunity with sbtype = auto to provide reasonable period "sane" SB types (especially critical to get the volume ramp correct), 

I don't think you can fix that with defaulting. Games that were specifically designed for SB16 and not for the lowest common denominator SB Pro 2 started appearing from 1995 onwards only. In 1993 the SB16 was a flop at release it was so buggy, plus most people had their SB Pros already, so adoption was slow. 

All in all, we can default to SB Pro 2 for sure, I think that's a better option. But that means we never auto select the SB16. We can't make that decision based on the configured graphics adapter/machine type (we cannot infer that "this is a 1995+ game").

Reply:
Also, if we had total freedom, I'd do this:

```ini
machine = pc | pcjr | tandy
video_adapter = hercules | cga | cga_mono | ega | svga_*
```

Or even split the `pc` type further if we want to emulate the lack of high DMA for `xt`:

```ini
machine = xt | at | pcjr | tandy
```

Certain `video_adapter` types would not be allowed for `pcjr` and `tandy`. Actually, we *could* make this scheme even work in a backwards-compatible manner if we wanted to!

Then technically you could put an EGA or VGA adapter in a Tandy. But maybe not that interesting/useful from a DOSBox perspective. But with the current conflated way of things you'd need something silly such as `tandy_ega` and `tandy_vga`...

I guess my point is the conflation of machines types with video adapters is bad enough already; I don't really want to pour oil on the fire... And _really_ dislike the conflation of vaguely (un)related things 😄

Reply:
The new functionality does what the user requests:

- If TandySound is running followed by a request for the SB16 using `'hdma' > 4` or GUS using `'gusdma' > 4, TandySound is shutdown to avoid IO port conflicts with the (now required) secondary DMA controller.

- Likewise, if SB16 or GUS are using the secondary controller followed by a request for TandySound, then the secondary controller is shutdown (hich cascades to shutting down devices on its channels).

- The same conflict resolution happens for DMA channels: The prior device is evicted to make way for the new resevation. For example, TandySound uses DMA channel 1 for its DAC, just like the Sound Blaster. Therefore, these two cannot co-exist on channel 1.

Reply:
Very nice @kcgen.